blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/ArtifactContainerImpl.java

```diff
@@ -1,122 +1,47 @@
         this.run = r;
         this.self = parent.getLink().rel("artifacts");
     }
 
-
     @Override
     public Link getLink() {
         return self;
     }
 
     @Override
     public BlueArtifact get(final String name) {
-        // Check security for artifacts
-        if(Functions.isArtifactsPermissionEnabled() && !run.hasPermission(Run.ARTIFACTS)) {
-            return null;
-        }
-
-        final VirtualFile file = run.getArtifactManager().root().child(name);
-        try {
-            if(file == null || !file.exists() || !file.isFile()) {
-                return null;
-            }
-        } catch (IOException e) {
-            throw new ServiceException.UnexpectedErrorException("Something is wrong with an artifact",e);
-        }
-        return new BlueArtifact() {
-            @Override
-            public String getName() {
-                return file.getName();
-            }
-
-            @Override
-            public String getPath() {
-                return name;
-            }
-
-            @Override
-            public String getUrl() {
-                return Stapler.getCurrentRequest().getContextPath() +
-                    "/" + run.getUrl()+"artifact/"+ name;
-            }
-
-            @Override
-            public long getSize() {
-                try {
-                    return file.length();
-                } catch (IOException e) {
-                    throw new ServiceException.UnexpectedErrorException("qError getting file length", e);
-                }
-            }
-
-            @Override
-            public Link getLink() {
-                return new Link(getUrl());
-            }
-        };
-    }
-
-    @Override
-    public Iterator<BlueArtifact> iterator(int start, int limit) {
-        // Check security for artifacts
-        if(Functions.isArtifactsPermissionEnabled() && !run.hasPermission(Run.ARTIFACTS)) {
-           return Iterators.emptyIterator();
-        }
-
-        List<Run.Artifact> artifactsUpTo = run.getArtifactsUpTo(limit);
-
-        // If start exceeds number of artifacts return an emtpy one.
-        if(start >= artifactsUpTo.size()) {
-            return Iterators.emptyIterator();
-        }
-
-        int calculatedLimit = limit;
-
-        if(calculatedLimit > artifactsUpTo.size()) {
-            calculatedLimit = artifactsUpTo.size();
-        }
-
-        List<Run.Artifact> artifacts = artifactsUpTo.subList(start, calculatedLimit);
-
-        return Iterators.transform(artifacts.iterator(), new Function<Run.Artifact, BlueArtifact>() {
+        return Iterators.find(iterator(), new Predicate<BlueArtifact>() {
             @Override
-            public BlueArtifact apply(@Nullable Run.Artifact artifact) {
-                if(artifact == null) {
-                    return null;
-                }
-
-                return new ArtifactImpl(run, artifact, ArtifactContainerImpl.this);
+            public boolean apply(@Nullable BlueArtifact input) {
```

Comment:
Artifact permission check `if(Functions.isArtifactsPermissionEnabled() && !run.hasPermission(Run.ARTIFACTS)) {}` is removed. 🐛  

Also, now that it returns ArtifactImpl or any implementation of BlueArtifact, `BlueArtifact.getSize()` is not implemented any more. Current functionality shows the  size of download.